x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/writer/ScheduledEventsWriter.java

```diff
@@ -58,4 +58,21 @@ public void write() throws IOException {
             ++eventIndex;
         }
     }
+
+    public void writeJson() throws IOException {
+        buffer.append('[');
+        boolean first = true;
+        for (ScheduledEvent event : scheduledEvents) {
+            if (first) {
+                first = false;
+            } else {
+                buffer.append(',');
+            }
+
+            try (XContentBuilder contentBuilder = XContentFactory.jsonBuilder()) {
+                buffer.append(Strings.toString(event.writeJson(bucketSpan, contentBuilder)));
+            }
+        }
+        buffer.append(']');
+    }
```

Comment:
It's very non-standard in Elasticsearch to have a JSON document whose outermost level is an array.  That's why you've had to hand-craft a writer instead of using standard mechanisms.  I think instead you should use a format similar to the [get scheduled events API response](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-get-calendar-event.html#ml-get-calendar-event-example), with the document being an object at its outermost level, and containing a field `events` that is the array of scheduled events.  If you can tolerate the superfluous `count` you could actually reuse `GetCalendarEventsAction.Response` to create and print this.  Or create a new very simple `ToXContentObject` that contains a `List<ScheduledEvent>` and uses `builder.field("events", scheduledEvents)` in its `toXContent` method.

Then you'll need to change the C++ parser to accept the extra object layer and `events` field, but there needs to be a second PR on the C++ side anyway and since the current C++ code is disabled this PR can still be merged before the second round of C++ updates.